File: evo_motion/view/specular.py
from os.path import join
from typing import Final, List, Tuple

import numpy as np
from glm import mat4, vec3, vec4
from OpenGL.GL import GL_TRIANGLES

from .constants import BYTES_PER_FLOAT, SHADER_ROOT_PATH
from .drawable import Drawable
from .program import Program


class SpecularObj(Drawable):
    __POSITION_SIZE: Final[int] = 3
    __NORMAL_SIZE: Final[int] = 3
    __STRIDE: Final[int] = (__POSITION_SIZE + __NORMAL_SIZE) * BYTES_PER_FLOAT

    def __init__(
        self,
        vertices: List[Tuple[float, float, float]],
        normals: List[Tuple[float, float, float]],
        ambient_color: vec4,
        diffuse_color: vec4,
        specular_color: vec4,
        shininess: float,
    ) -> None:
        assert len(vertices) == len(normals), (
            f"vertices number does not match normals number : "
            f"{len(vertices)} != {len(normals)}"
        )

        vbo_data = np.array(
            [
                d
                for v, n in zip(vertices, normals)
                for d in [v[0], v[1], v[2], n[0], n[1], n[2]]
            ],
            dtype=np.float32,
        )

        self.__nb_vertices = len(vertices)

        self.__program: Program = (
            Program.Builder(
                join(SHADER_ROOT_PATH, "specular_vs.glsl"),
                join(SHADER_ROOT_PATH, "specular_fs.glsl"),
            )
            .add_uniform_mat_4fv("u_mvp_matrix")
            .add_uniform_mat_4fv("u_mv_matrix")
            .add_uniform_4fv("u_ambient_color")
            .add_uniform_4fv("u_diffuse_color")
            .add_uniform_4fv("u_specular_color")
            .add_uniform_3fv("u_light_pos")
            .add_uniform_1f("u_distance_coef")
            .add_uniform_1f("u_light_coef")
            .add_uniform_1f("u_shininess")
            .add_uniform_3fv("u_cam_pos")
            .new_buffer("vertices_normals_buffer", vbo_data)
            .add_attrib("a_position")
            .add_attrib("a_normal")
            .build_buffer()
            .build_program()
        )

        self.__ambient_color = ambient_color
        self.__diffuse_color = diffuse_color
        self.__specular_color = specular_color
        self.__shininess = shininess

    def draw(
        self,
        mvp_matrix: mat4,
        mv_matrix: mat4,
        light_pos_from_camera: vec3,
        camera_pos: vec3,
    ) -> None:
        self.__program.use()

        self.__program.attrib(
            "vertices_normals_buffer",
            "a_position",
            SpecularObj.__POSITION_SIZE,
            SpecularObj.__STRIDE,
            0,
        )

        self.__program.attrib(
            "vertices_normals_buffer",
            "a_normal",
            SpecularObj.__NORMAL_SIZE,
            SpecularObj.__STRIDE,
            SpecularObj.__POSITION_SIZE * BYTES_PER_FLOAT,
        )

        self.__program.uniform_mat_4fv("u_mvp_matrix", mvp_matrix)
        self.__program.uniform_mat_4fv("u_mv_matrix", mv_matrix)

        self.__program.uniform_3fv("u_light_pos", light_pos_from_camera)
        self.__program.uniform_3fv("u_cam_pos", camera_pos)

        self.__program.uniform_4fv("u_ambient_color", self.__ambient_color)
        self.__program.uniform_4fv("u_diffuse_color", self.__diffuse_color)
        self.__program.uniform_4fv("u_specular_color", self.__specular_color)

        self.__program.uniform_1f("u_distance_coef", 0.0)
        self.__program.uniform_1f("u_light_coef", 1.0)
        self.__program.uniform_1f("u_shininess", self.__shininess)

        Program.draw_arrays(GL_TRIANGLES, 0, self.__nb_vertices)

        self.__program.disable_attribs("vertices_normals_buffer")

File: src/view/specular.cpp
//
// Created by samuel on 17/12/22.
//

#include "constants.h"
#include "specular.h"

OBjSpecular::OBjSpecular(
    const std::vector<std::tuple<float, float, float>> &vertices,
    const std::vector<std::tuple<float, float, float>> &normals,
    glm::vec4 ambient_color,
    glm::vec4 diffuse_color,
    glm::vec4 specular_color, float shininess)
    : ambient_color(ambient_color),
      diffuse_color(diffuse_color),
      specular_color(specular_color),
      nb_vertices(int(vertices.size())),
      shininess(shininess) {

    std::vector<float> vbo_data;
    for (int i = 0; i < vertices.size(); i++) {
        vbo_data.push_back(std::get<0>(vertices[i]));
        vbo_data.push_back(std::get<1>(vertices[i]));
        vbo_data.push_back(std::get<2>(vertices[i]));

        vbo_data.push_back(std::get<0>(normals[i]));
        vbo_data.push_back(std::get<1>(normals[i]));
        vbo_data.push_back(std::get<2>(normals[i]));
    }

    program = Program::Builder(
        "./shaders/specular_vs.glsl",
        "./shaders/specular_fs.glsl"
    ).add_uniform("u_mvp_matrix")
        .add_uniform("u_mv_matrix")
        .add_uniform("u_ambient_color")
        .add_uniform("u_diffuse_color")
        .add_uniform("u_specular_color")
        .add_uniform("u_light_pos")
        .add_uniform("u_distance_coef")
        .add_uniform("u_light_coef")
        .add_uniform("u_shininess")
        .add_uniform("u_cam_pos")
        .add_buffer("vertices_normals_buffer", vbo_data)
        .add_attribute("a_position")
        .add_attribute("a_normal")
        .build();
}

void
OBjSpecular::draw(glm::mat4 mvp_matrix, glm::mat4 mv_matrix,
                  glm::vec3 light_pos_from_camera, glm::vec3 camera_pos) {
    program.use();

    program.attrib(
        "a_position",
        "vertices_normals_buffer",
        POSITION_SIZE,
        STRIDE,
        0
    );

    program.attrib(
        "a_normal",
        "vertices_normals_buffer",
        NORMAL_SIZE,
        STRIDE,
        POSITION_SIZE * BYTES_PER_FLOAT
    );

    program.uniform_mat4("u_mvp_matrix", mvp_matrix);
    program.uniform_mat4("u_mv_matrix", mv_matrix);

    program.uniform_vec3("u_light_pos", light_pos_from_camera);
    program.uniform_vec3("u_cam_pos", camera_pos);

    program.uniform_vec4("u_ambient_color", ambient_color);
    program.uniform_vec4("u_diffuse_color", diffuse_color);
    program.uniform_vec4("u_specular_color", specular_color);

    program.uniform_float("u_distance_coef", 0.f);
    program.uniform_float("u_light_coef", 1.f);
    program.uniform_float("u_shininess", shininess);

    Program::draw_arrays(GL_TRIANGLES, 0, nb_vertices);

    program.disable_attrib_array();
}

OBjSpecular::~OBjSpecular() {
    program.kill();
}
